[Site isolation] Save As PDF does not render site isolated iframes

mattwoodrow opened the pull request:
#### dace43e032d86999d6d06c321bd9fe671298166e
<pre>
[Site isolation] Save As PDF does not render site isolated iframes
<a href="https://bugs.webkit.org/show_bug.cgi?id=282664">https://bugs.webkit.org/show_bug.cgi?id=282664</a>
<a href="https://rdar.apple.com/139325001">rdar://139325001</a>

Reviewed by Kimmo Kinnunen.

Changes the DrawRemoteToPDF triangle IPC (UI -&gt; WCP -&gt; GPUP -&gt; UI) into two IPC
calls with async reply:

WebPage::DrawToSnapshot(UI &lt;-&gt; WCP)  which builds a cross-process snapshot (as a
set of DisplayLists) in the GPUP and returns once this is fully available.

GPUProcess::sinkSnapshotToPDF (UI &lt;-&gt; GPUP) which consumes a snapshot object and
converts it into a SharedBuffer to return. We have the option to add other
versions of this which sink into a bitmap instead.

Using these lets us take advantage of the existing async reply handlers, where a
process crash will cause the handlers to be resolved. This should mean that we
always resolve the pending PDF generation, even if child process crashes.

Uses the newly available RemoteDisplayListRecorderProxy, and creates a
snapshotting specific subclass RemoteSnapshotRecorderProxy.  This directly
generates GPUP display lists, (including recording placeholder items for remote
subframes) while drawing the WCP content, and then a new API
sinkSnapshotRecorderIntoSnapshotFrame to transfer that display list into the
GPU&apos;s snapshot object.

When painting encounters a remote frame it requests that frame to also create a
GPUP display list to be included into the same snapshot. The outer frame records
a DrawPlaceholder command as a placeholder for the inner frame&apos;s display list.
This is only exposed on the snapshot recorder, since it has no meaning for a
generic GraphicsContext.

All functions return an async callback, and a new callback aggregator is
introduced so that we don&apos;t resolve back to the UI process until all involved
frames have had a confirmed response from the GPUP.

On the GPUP side, we store all the accumulated display lists in a
&apos;RemoteSnapshot&apos; object, accessed from the main thread. The DisplayList type
isn&apos;t properly threadsafe, but should work correctly since the causes of races
(like mutating Font or Gradient) only happen in the web process, never in the
GPUP versions of these. We do unfortunately have to track the work queue that
these display lists were created on, so that we can dispatch them back for
destruction. A truly threadsafe DisplayList would simplify this code immensely.

Removed old crashtest that was specific to the old WebProcess -&gt; GPU
drawRemoteToPDF (and ImageBufferDisplayListBackend usage), which isn&apos;t possible
with the new API.

* Source/WTF/wtf/CallbackAggregator.h:
(WTF::SuccessCallbackAggregatorOnThread::create):
(WTF::SuccessCallbackAggregatorOnThread::~SuccessCallbackAggregatorOnThread):
(WTF::SuccessCallbackAggregatorOnThread::failed):
(WTF::SuccessCallbackAggregatorOnThread::CompletionHandler&lt;void):
(WTF::SuccessCallbackAggregatorOnThread::SuccessCallbackAggregatorOnThread):
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/RemoteFrameClient.h:
* Source/WebCore/page/RemoteFrameView.cpp:
(WebCore::RemoteFrameView::paintContents):
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::DrawPlaceholder::DrawPlaceholder):
(WebCore::DisplayList::DrawPlaceholder::apply const):
(WebCore::DisplayList::DrawPlaceholder::dump const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::drawPlaceholder):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::getOrCreateSnapshot):
(WebKit::GPUProcess::sinkCompletedSnapshotToPDF):
(WebKit::GPUProcess::releaseSnapshot):
(WebKit::GPUProcess::didDrawRemoteToPDF): Deleted.
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in: Copied from Source/WebKit/UIProcess/GPU/GPUProcessProxy.messages.in.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::createSnapshotRecorder):
(WebKit::RemoteRenderingBackend::sinkSnapshotRecorderIntoSnapshotFrame):
(WebKit::RemoteRenderingBackend::didDrawRemoteToPDF): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteSnapshot.cpp: Added.
(WebKit::RemoteSnapshot::create):
(WebKit::RemoteSnapshot::addFrameReference):
(WebKit::RemoteSnapshot::setFrame):
(WebKit::RemoteSnapshot::applyFrame const):
(WebKit::RemoteSnapshot::isComplete const):
(WebKit::RemoteSnapshot::DisplayListAndReleaseDispatcher::DisplayListAndReleaseDispatcher):
(WebKit::RemoteSnapshot::DisplayListAndReleaseDispatcher::~DisplayListAndReleaseDispatcher):
(WebKit::RemoteSnapshot::drawToPDF):
* Source/WebKit/GPUProcess/graphics/RemoteSnapshot.h: Added.
* Source/WebKit/GPUProcess/graphics/RemoteSnapshotDisplayListIdentifier.h: Added.
* Source/WebKit/GPUProcess/graphics/RemoteSnapshotIdentifier.h: Added.
* Source/WebKit/GPUProcess/graphics/RemoteSnapshotRecorder.cpp: Added.
(WebKit::RemoteSnapshotRecorder::create):
(WebKit::RemoteSnapshotRecorder::RemoteSnapshotRecorder):
(WebKit::RemoteSnapshotRecorder::startListeningForIPC):
(WebKit::RemoteSnapshotRecorder::stopListeningForIPC):
(WebKit::RemoteSnapshotRecorder::snapshot const):
(WebKit::RemoteSnapshotRecorder::drawSnapshotFrame):
* Source/WebKit/GPUProcess/graphics/RemoteSnapshotRecorder.h: Added.
* Source/WebKit/GPUProcess/graphics/RemoteSnapshotRecorder.messages.in: Copied from Source/WebKit/UIProcess/GPU/GPUProcessProxy.messages.in.
* Source/WebKit/GPUProcess/graphics/RemoteSnapshotRecorderIdentifier.h: Added.
* Source/WebKit/GPUProcess/webrtc/RemoteAudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit::RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit):
* Source/WebKit/Platform/IPC/ObjectIdentifierReference.serialization.in:
* Source/WebKit/Scripts/webkit/messages.py:
(atomic_object_identifier):
(serialized_identifiers):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView createPDFWithConfiguration:completionHandler:]):
* Source/WebKit/UIProcess/Cocoa/GPUProcessProxyCocoa.mm:
(WebKit::GPUProcessProxy::sinkCompletedSnapshotToPDF):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::releaseSnapshot):
(WebKit::GPUProcessProxy::didDrawRemoteToPDF): Deleted.
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::drawToPDF):
(WebKit::WebPageProxy::drawFrameToSnapshot):
(WebKit::WebPageProxy::drawRemoteToPDF): Deleted.
(WebKit::WebPageProxy::didDrawRemoteToPDF): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.cpp:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::createSnapshotRecorder):
(WebKit::RemoteRenderingBackendProxy::sinkSnapshotRecorderIntoSnapshotFrame):
(WebKit::RemoteRenderingBackendProxy::didDrawRemoteToPDF): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteSnapshotRecorderProxy.cpp: Added.
(WebKit::RemoteSnapshotRecorderProxy::RemoteSnapshotRecorderProxy):
(WebKit::RemoteSnapshotRecorderProxy::drawSnapshotFrame):
* Source/WebKit/WebProcess/GPU/graphics/RemoteSnapshotRecorderProxy.h: Added.
(WebKit::RemoteSnapshotRecorderProxy::identifier const):
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::query):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::paintContents):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_toolbarsAreVisible):
(WebKit::WebPage::~WebPage):
(WebKit::WebPage::executeEditingCommand):
(WebKit::WebPage::dumpHistoryForTesting):
(WebKit::WebPage::tryMarkLayersVolatile):
(WebKit::WebPage::updateIsInWindow):
(WebKit::WebPage::setActivityState):
(WebKit::WebPage::updatePreferences):
(WebKit::WebPage::drawToPDF):
(WebKit::WebPage::paintRemoteFrameContents):
(WebKit::WebPage::drawToSnapshot):
(WebKit::WebPage::drawFrameToSnapshot):
(WebKit::WebPage::insertTextAsync):
(WebKit::WebPage::didCommitLoad):
(WebKit::WebPage::updateWebsitePolicies):
(WebKit::WebPage::setIsNavigatingToAppBoundDomain):
(WebKit::WebPage::drawRemoteToPDF): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/300358@main">https://commits.webkit.org/300358@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/50d9c1a075b6720750fd111d07048dd4150bb742

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122250 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41953 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32622 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128827 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/74344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d4132dd0-4a33-4dfc-9810-1f3296f8a679) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124126 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42667 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50546 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92922 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/74344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6cef9893-5cd2-46f4-b720-04396a024b2e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125202 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34032 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109464 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73576 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6fa215a7-96be-41f5-8373-dc65a733874d) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/121607 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33041 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27627 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72316 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/114398 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103540 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27820 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131574 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/120776 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49189 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37424 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101494 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49564 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105675 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101364 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25705 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46722 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24840 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/45943 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49046 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54781 "Failed to build and analyze WebKit") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/150935 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/48516 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38618 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51866 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50196 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/60676ffc0f89eea37b3f983871a525f689e9db28. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50643

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/50d9c1a075b6720750fd111d07048dd4150bb742. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50643

mattwoodrow committed:
[Site isolation] Save As PDF does not render site isolated iframes
https://bugs.webkit.org/show_bug.cgi?id=282664
rdar://139325001

Reviewed by Kimmo Kinnunen.

Changes the DrawRemoteToPDF triangle IPC (UI -> WCP -> GPUP -> UI) into two IPC
calls with async reply:

WebPage::DrawToSnapshot(UI <-> WCP)  which builds a cross-process snapshot (as a
set of DisplayLists) in the GPUP and returns once this is fully available.

GPUProcess::sinkSnapshotToPDF (UI <-> GPUP) which consumes a snapshot object and
converts it into a SharedBuffer to return. We have the option to add other
versions of this which sink into a bitmap instead.

Using these lets us take advantage of the existing async reply handlers, where a
process crash will cause the handlers to be resolved. This should mean that we
always resolve the pending PDF generation, even if child process crashes.

Uses the newly available RemoteDisplayListRecorderProxy, and creates a
snapshotting specific subclass RemoteSnapshotRecorderProxy.  This directly
generates GPUP display lists, (including recording placeholder items for remote
subframes) while drawing the WCP content, and then a new API
sinkSnapshotRecorderIntoSnapshotFrame to transfer that display list into the
GPU's snapshot object.

When painting encounters a remote frame it requests that frame to also create a
GPUP display list to be included into the same snapshot. The outer frame records
a DrawPlaceholder command as a placeholder for the inner frame's display list.
This is only exposed on the snapshot recorder, since it has no meaning for a
generic GraphicsContext.

All functions return an async callback, and a new callback aggregator is
introduced so that we don't resolve back to the UI process until all involved
frames have had a confirmed response from the GPUP.

On the GPUP side, we store all the accumulated display lists in a
'RemoteSnapshot' object, accessed from the main thread. The DisplayList type
isn't properly threadsafe, but should work correctly since the causes of races
(like mutating Font or Gradient) only happen in the web process, never in the
GPUP versions of these. We do unfortunately have to track the work queue that
these display lists were created on, so that we can dispatch them back for
destruction. A truly threadsafe DisplayList would simplify this code immensely.

Removed old crashtest that was specific to the old WebProcess -> GPU
drawRemoteToPDF (and ImageBufferDisplayListBackend usage), which isn't possible
with the new API.

* Source/WTF/wtf/CallbackAggregator.h:
(WTF::SuccessCallbackAggregatorOnThread::create):
(WTF::SuccessCallbackAggregatorOnThread::~SuccessCallbackAggregatorOnThread):
(WTF::SuccessCallbackAggregatorOnThread::failed):
(WTF::SuccessCallbackAggregatorOnThread::CompletionHandler<void):
(WTF::SuccessCallbackAggregatorOnThread::SuccessCallbackAggregatorOnThread):
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/RemoteFrameClient.h:
* Source/WebCore/page/RemoteFrameView.cpp:
(WebCore::RemoteFrameView::paintContents):
* Source/WebCore/platform/graphics/displaylists/DisplayListItem.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.cpp:
(WebCore::DisplayList::DrawPlaceholder::DrawPlaceholder):
(WebCore::DisplayList::DrawPlaceholder::apply const):
(WebCore::DisplayList::DrawPlaceholder::dump const):
* Source/WebCore/platform/graphics/displaylists/DisplayListItems.h:
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.cpp:
(WebCore::DisplayList::RecorderImpl::drawPlaceholder):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorderImpl.h:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::getOrCreateSnapshot):
(WebKit::GPUProcess::sinkCompletedSnapshotToPDF):
(WebKit::GPUProcess::releaseSnapshot):
(WebKit::GPUProcess::didDrawRemoteToPDF): Deleted.
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcess.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteDisplayListRecorder.messages.in: Copied from Source/WebKit/UIProcess/GPU/GPUProcessProxy.messages.in.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContext.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::createSnapshotRecorder):
(WebKit::RemoteRenderingBackend::sinkSnapshotRecorderIntoSnapshotFrame):
(WebKit::RemoteRenderingBackend::didDrawRemoteToPDF): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteSnapshot.cpp: Added.
(WebKit::RemoteSnapshot::create):
(WebKit::RemoteSnapshot::addFrameReference):
(WebKit::RemoteSnapshot::setFrame):
(WebKit::RemoteSnapshot::applyFrame const):
(WebKit::RemoteSnapshot::isComplete const):
(WebKit::RemoteSnapshot::DisplayListAndReleaseDispatcher::DisplayListAndReleaseDispatcher):
(WebKit::RemoteSnapshot::DisplayListAndReleaseDispatcher::~DisplayListAndReleaseDispatcher):
(WebKit::RemoteSnapshot::drawToPDF):
* Source/WebKit/GPUProcess/graphics/RemoteSnapshot.h: Added.
* Source/WebKit/GPUProcess/graphics/RemoteSnapshotDisplayListIdentifier.h: Added.
* Source/WebKit/GPUProcess/graphics/RemoteSnapshotIdentifier.h: Added.
* Source/WebKit/GPUProcess/graphics/RemoteSnapshotRecorder.cpp: Added.
(WebKit::RemoteSnapshotRecorder::create):
(WebKit::RemoteSnapshotRecorder::RemoteSnapshotRecorder):
(WebKit::RemoteSnapshotRecorder::startListeningForIPC):
(WebKit::RemoteSnapshotRecorder::stopListeningForIPC):
(WebKit::RemoteSnapshotRecorder::snapshot const):
(WebKit::RemoteSnapshotRecorder::drawSnapshotFrame):
* Source/WebKit/GPUProcess/graphics/RemoteSnapshotRecorder.h: Added.
* Source/WebKit/GPUProcess/graphics/RemoteSnapshotRecorder.messages.in: Copied from Source/WebKit/UIProcess/GPU/GPUProcessProxy.messages.in.
* Source/WebKit/GPUProcess/graphics/RemoteSnapshotRecorderIdentifier.h: Added.
* Source/WebKit/GPUProcess/webrtc/RemoteAudioMediaStreamTrackRendererInternalUnitManager.cpp:
(WebKit::RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit::RemoteAudioMediaStreamTrackRendererInternalUnitManagerUnit):
* Source/WebKit/Platform/IPC/ObjectIdentifierReference.serialization.in:
* Source/WebKit/Scripts/webkit/messages.py:
(atomic_object_identifier):
(serialized_identifiers):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView createPDFWithConfiguration:completionHandler:]):
* Source/WebKit/UIProcess/Cocoa/GPUProcessProxyCocoa.mm:
(WebKit::GPUProcessProxy::sinkCompletedSnapshotToPDF):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::releaseSnapshot):
(WebKit::GPUProcessProxy::didDrawRemoteToPDF): Deleted.
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::drawToPDF):
(WebKit::WebPageProxy::drawFrameToSnapshot):
(WebKit::WebPageProxy::drawRemoteToPDF): Deleted.
(WebKit::WebPageProxy::didDrawRemoteToPDF): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.cpp:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::createSnapshotRecorder):
(WebKit::RemoteRenderingBackendProxy::sinkSnapshotRecorderIntoSnapshotFrame):
(WebKit::RemoteRenderingBackendProxy::didDrawRemoteToPDF): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteSnapshotRecorderProxy.cpp: Added.
(WebKit::RemoteSnapshotRecorderProxy::RemoteSnapshotRecorderProxy):
(WebKit::RemoteSnapshotRecorderProxy::drawSnapshotFrame):
* Source/WebKit/WebProcess/GPU/graphics/RemoteSnapshotRecorderProxy.h: Added.
(WebKit::RemoteSnapshotRecorderProxy::identifier const):
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::query):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::paintContents):
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_toolbarsAreVisible):
(WebKit::WebPage::~WebPage):
(WebKit::WebPage::executeEditingCommand):
(WebKit::WebPage::dumpHistoryForTesting):
(WebKit::WebPage::tryMarkLayersVolatile):
(WebKit::WebPage::updateIsInWindow):
(WebKit::WebPage::setActivityState):
(WebKit::WebPage::updatePreferences):
(WebKit::WebPage::drawToPDF):
(WebKit::WebPage::paintRemoteFrameContents):
(WebKit::WebPage::drawToSnapshot):
(WebKit::WebPage::drawFrameToSnapshot):
(WebKit::WebPage::insertTextAsync):
(WebKit::WebPage::didCommitLoad):
(WebKit::WebPage::updateWebsitePolicies):
(WebKit::WebPage::setIsNavigatingToAppBoundDomain):
(WebKit::WebPage::drawRemoteToPDF): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/300358@main
